include/nbl/video/IDeviceMemoryAllocator.h

```diff
@@ -27,7 +28,93 @@ class IDeviceMemoryAllocator
 		// size_t opaqueCaptureAddress = 0u; Note that this mechanism is intended only to support capture/replay tools, and is not recommended for use in other applications.
 	};
 
+	//! IMemoryTypeIterator extracts memoryType indices from memoryTypeMask in arbitrary order
+	//! which is used to give priority to memoryTypes in try-allocate usages where allocations may fail with some memoryTypes
+	//! IMemoryTypeIterator will construct SAllocateInfo from object's memory requirements, allocateFlags and dedication using operator()
+	class IMemoryTypeIterator
+	{
+	public:
+		IMemoryTypeIterator(const IDriverMemoryBacked::SDriverMemoryRequirements2& reqs, core::bitflag<IDriverMemoryAllocation::E_MEMORY_ALLOCATE_FLAGS> allocateFlags)
+			: m_allocateFlags(static_cast<uint32_t>(allocateFlags.value))
+			, m_reqs(reqs)
+		{}
+
+		static inline uint32_t end() {return 32u;}
+
+		IMemoryTypeIterator& operator++()
+		{
+			advance();
+			return *this;
+		}
+
+		inline SAllocateInfo operator()(IDriverMemoryBacked* dedication)
+		{
+			SAllocateInfo ret;
+			ret.size = m_reqs.size;
+			ret.flags = m_allocateFlags;
+			ret.memoryTypeIndex = dereference();
+			ret.dedication = dedication;
+			return ret;
+		}
+
+	protected:
+		virtual uint32_t dereference() const = 0;
+		virtual void advance() = 0;
+		virtual bool operator==(uint32_t rhs) const = 0;
+		virtual bool operator!=(uint32_t rhs) const = 0;
```

Comment:
what do you need those for?

Reply:
easiest way I could find to != IMemoryTypeItr::end() which is a uint32_t
If ::end it's iterator type instead (like usual iterators) then it'll also need a operator +

Reply:
I dont think these need to be virtual you could inline them and compare against `dereference()`

Reply:
oh yea :D
thanks for the tip